Sources/Algorithms/Stride.swift

```diff
@@ -0,0 +1,160 @@
+  /// O(_k_), where _k_ is the striding `step`.
+  ///
+  /// - Parameter step: The amount to step with each iteration.
+  /// - Returns: Returns a collection stepping through the elements by the
+  /// specified amount.
+  public func striding(by step: Int) -> Stride<Self> {
+    Stride(base: self, stride: step)
+  }
+}
+
+public struct Stride<Base: Collection> {
+  
+  public let base: Base
+  public let stride: Int
+  
+  init(base: Base, stride: Int) {
+    precondition(stride > 0, "striding must be greater than zero")
+    self.base = base
+    self.stride = stride
+  }
+}
+
+extension Stride {
+  public func striding(by step: Int) -> Self {
+    Stride(base: base, stride: stride * step)
+  }
+}
+
+extension Stride: Collection {
+  
+  public struct Index: Comparable {
+    
+    let base: Base.Index
+    
+    init(_ base: Base.Index) {
+      self.base = base
+    }
+    
+    public static func < (lhs: Index, rhs: Index) -> Bool {
+      lhs.base < rhs.base
+    }
+  }
+  
+  public var startIndex: Index {
+    Index(base.startIndex)
+  }
+  
+  public var endIndex: Index {
+    Index(base.endIndex)
+  }
+  
+  public subscript(i: Index) -> Base.Element {
+    base[i.base]
+  }
+  
+  public func index(after i: Index) -> Index {
+    precondition(i.base < base.endIndex, "Advancing past end index")
+    return base.index(i.base, offsetBy: stride, limitedBy: base.endIndex)
+      .map(Index.init) ?? endIndex
+  }
+  
+  public func index(
+    _ i: Index,
+    offsetBy distance: Int,
+    limitedBy limit: Index
+  ) -> Index? {
+    base.index(i.base, offsetBy: distance * stride, limitedBy: limit.base)
+      .map(Index.init)
+  }
+  
+  public var count: Int {
+    (base.count - 1) / stride + 1
+  }
+  
+  public func distance(from start: Index, to end: Index) -> Int {
+    base.distance(from: start.base, to: end.base) / stride
+  }
+  
+  public func index(_ i: Index, offsetBy distance: Int) -> Index {
+    Index(base.index(i.base, offsetBy: distance * stride))
+  }
+
+}
+
+extension Stride: BidirectionalCollection
+  where Base: RandomAccessCollection {
+  
+  public func index(before i: Index) -> Index {
+    precondition(i.base > base.startIndex, "Incrementing past start index")
+    if i == endIndex {
+      let count = base.count
+      precondition(count > 0, "Can't move before the starting index")
+      return Index(
+        base.index(base.endIndex, offsetBy: -((count - 1) % stride + 1))
+      )
+    } else {
+      guard let step = base.index(
+        i.base,
+        offsetBy: -stride,
+        limitedBy: startIndex.base
+      ) else {
+        fatalError("Incrementing past start index")
+      }
+      return Index(step)
+    }
+  }
+}
+
+extension Stride: RandomAccessCollection
+  where Base: RandomAccessCollection {}
+
+extension Stride: Equatable
+  where Base.Element: Equatable {
+  
+  public static func == (lhs: Stride, rhs: Stride) -> Bool {
+    guard lhs.count == rhs.count else { return false }
```

Comment:
You can leave this check out — `elementsEqual` will fail if they have different lengths.